### PR TITLE
cmake change for correct placement of autodetected tune & arch flags of GCC and SVE enablement on AARCH64

### PR DESCRIPTION
**Pull request by abondarev84**
also updated README for SVE short explanation

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -165,26 +165,48 @@ if (CMAKE_COMPILER_IS_GNUCC AND NOT CROSS_COMPILE)
...
     string(REGEX REPLACE "${ARCH_FLAG}=[ \t]*([^ \n]*)[ \n].*" "\\1" GNUCC_ARCH "${_GCC_OUTPUT}")
 
+    string(FIND "${_GCC_OUTPUT_TUNE}" "${TUNE_FLAG}=" POS_TUNE)
+    string(SUBSTRING "${_GCC_OUTPUT_TUNE}" ${POS_TUNE} -1 _GCC_OUTPUT_TUNE)
+    string(REGEX REPLACE "${TUNE_FLAG}=[ \t]*([^ \n]*)[ \n].*" "\\1" GNUCC_TUNE "${_GCC_OUTPUT_TUNE}")
+
+    string(FIND "${GNUCC_ARCH}" "sve" POS_SVE)
+    string(FIND "${GNUCC_ARCH}" "sve2" POS_SVE2)
+    string(FIND "${GNUCC_ARCH}" "sve2-bitperm" POS_SVE2_BITPERM)
+    if (NOT POS_SVE EQUAL 0)
```
Comment (markos) [inline CMakeLists.txt:189]: SVE* should not be enabled if detected, BUILD_SVE* should only be enabled by the user explicitly.

File: CMakeLists.txt [hunk 2]
```diff
@@ -225,23 +247,12 @@ if (ARCH_IA32 OR ARCH_X86_64)
     endif()
 endif()
 
-if (ARCH_AARCH64)
-    if (BUILD_SVE2_BITPERM)
-        set(GNUCC_ARCH "${GNUCC_ARCH}+sve2-bitperm")
```
Comment (markos) [inline CMakeLists.txt:230]: this is actually a more proper way to to it, the reason it fails right now is because it adds sve even if it was already in the GNUCC_ARCH variable.

Comment (abondarev84) [reply CMakeLists.txt:230]: the end of autodetect flags for arch was `+nodotprod+sve`, by adding something after it, gcc would error out saying
```
Assembler messages:
Error: must specify extensions to add before specifying those to remove
```

possibly autodetect of tune/arch flags should be reconsidered in total?

Comment (markos) [reply CMakeLists.txt:230]: I agree that it should be reworked, but this is a low priority right now. , For that matter, I would also like to change build system altogether, but this is even lower priority. At the moment we have to adapt to the existing system. So, the goal is to detect is SVE* is available, but only actually enable it if BUILD_SVE* variables are set to true. I wonder if we should also allow building SVE on non-SVE systems -eg. a distro builder that lacks SVE, should be able to build the binaries.

**Comment by markos** [comment]
@abondarev84 By the way, when you are done, could you please submit this PR against develop branch? Thanks.

**Comment by abondarev84** [comment]
> 

of course

File: CMakeLists.txt (hunk 2 above)
Comment (abondarev84) [reply CMakeLists.txt:230]: Interesting idea.
I believe it deserves its own attention.

File: CMakeLists.txt [hunk 3]
```diff
@@ -226,22 +248,26 @@ if (ARCH_IA32 OR ARCH_X86_64)
...
 
 if (ARCH_AARCH64)
-    if (BUILD_SVE2_BITPERM)
+    if (BUILD_SVE2_BITPERM AND NOT SVE2_BITPERM_FOUND)
         set(GNUCC_ARCH "${GNUCC_ARCH}+sve2-bitperm")
-    elseif (BUILD_SVE2)
+    elseif (BUILD_SVE2 AND NOT SVE2_FOUND)
         set(GNUCC_ARCH "${GNUCC_ARCH}+sve2")
-    elseif (BUILD_SVE)
+    elseif (BUILD_SVE AND NOT SVE_FOUND)
```
Comment (markos) [inline CMakeLists.txt:255]: I may be misunderstanding it here, If BUILD_SVE is set, and SVE is NOT found, then enable it? Should it not be the other way around?

**Comment by markos** [comment]
I confirm that it works on my graviton3 system with gcc, unfortunately it fails with clang (tested 11 and 13).

File: CMakeLists.txt (hunk 3 above)
Comment (abondarev84) [reply CMakeLists.txt:255]: you understood correctly, I'll explain the logic.  

since this section only adds the various sve flags. to not break the compilation, they will be added only if they don't exist currently in the arch field ${GNUCC_ARCH}.  
inside the arch folder there is a check if sve is actually supported on the system (this is in case one of the sve flags has been activated) and errors out if it doesn't (as per the original behavior).  
this way one of the SVE flags isn't turned on automatically but according to user request, and in that specific instance, it doesn't break the currently present arch field with an addition at the end of the line.

**Comment by abondarev84** [comment]
> I confirm that it works on my graviton3 system with gcc, unfortunately it fails with clang (tested 11 and 13).

do you mind referencing to what failed? run jenking run 124 is still running without errors so far.

for run 123 I see that sve wasn't found but it wasn't in the flags or cmake command.

**Comment by markos** [comment]
Jenkins jobs do not run SVE specific builds, it's only NEON. If you try building on graviton3 system with
```CC=clang CXX=clang++ cmake -DCMAKE_BUILD_TYPE=Release -DBUILD_SVE=on ..```

you will see the error.

File: CMakeLists.txt (hunk 3 above)
Comment (markos) [reply CMakeLists.txt:255]: It seems counter-intuitive. What if they exist in the GNUCC_ARCH but user explicitly sets BUILD_SVE=off? SVE Code paths will be disabled, but that does not mean the compiler cannot optimize the build using actual SVE instructions. Which may be confusing to the user, and it may lead to SIGILL on non-SVE systems. But I think that's the case also for AVX2/AVX512 on Intel. I'm afraid, the hardware features detection should be reworked entirely soon. In any case for this particular problem, could you please rephrase the logic so that it's more obvious why this is so?

**Comment by abondarev84** [comment]
> Jenkins jobs do not run SVE specific builds (yet), it's only NEON. If you try building on graviton3 system with `CC=clang CXX=clang++ cmake -DCMAKE_BUILD_TYPE=Release -DBUILD_SVE=on ..`
> 
> you will see the error.

there were no changes to clang logic path.  
what I'm seeing as the cause here is that the error of not having arm_sve.h happens for lack of a file identification in the includes.
I can update the commit message to reflect that these fixes were implemented for gcc.

File: CMakeLists.txt (hunk 3 above)
Comment (abondarev84) [reply CMakeLists.txt:255]: the current logic is 
if (not found and yes requested) 
    add flags
    
I'm not sure how to rephrase this as this is very specific condition and not hidden.

> but user explicitly sets BUILD_SVE=off? why would a user turn off something they didn't turn on? is there an example in the cmake for this logic somewhere?

Comment (markos) [reply CMakeLists.txt:255]: ok, what I mean is this: regardless if SVE found or not, it should be enabled only if BUILD_SVE is set to true, so the old logic was enough. In order to build for SVE, you only really need compiler support, not platform support (now running the tests is another matter ofc). But I now see what your method solves, you need the extra check so as not to add "+sve" a second time to the flags -which would break the build. Ok, this will work, but for the long term I will have to think of a better solution for all architectures, so you can expect this detection system to change in the future in general.

**Comment by markos** [comment]
The problem is that it has to work for clang as well, before it can be merged. I see that clang correctly sets -march=armv8+sve in the CFLAGS/CXXFLAGS, so I'm wondering why it fails to detect arm_sve.h.

**Comment by markos** [comment]
right, ARCH_C_FLAGS/ARCH_CXX_FLAGS are set correctly, but they are not used in CHECK_INCLUDE_FILE_CXX for clang, which leads to failure.

**Comment by markos** [comment]
yup, that's it, adding this line before the CHECK_INCLUDE_FILE_CXX, line 361, fixes the build:
```set(CMAKE_REQUIRED_FLAGS ${ARCH_CXX_FLAGS})```

**Comment by markos** [comment]
could you also remove Jenkinsfile from the PR? thanks

**Comment by abondarev84** [comment]
> could you also remove Jenkinsfile from the PR? thanks

I updated my fork, should I remove the merge from remote and leave my repo at my newest commit?
(I can't just remove the file from the pr in githubs ui it seems)

**Comment by abondarev84** [comment]
@markos removed the merge from your origin.

**Comment by markos** [comment]
ok, let's fix the clang issue as mentioned in https://github.com/VectorCamp/vectorscan/pull/125#issuecomment-1250329402 and wait for jenkins to be done and it should be good to go :)
